### PR TITLE
Improve CertAuthorityWatcher

### PR DESCRIPTION
**Pull request by rosstimothy**
CertAuthorityWatcher and its usage are refactored to allow for
all the following:
 - eliminate retransmission of the same CAs
 - reduce memory usage by having one local watcher per proxy
 - adds the ability to filter only the CAs that are desired
 - reduce the time required to send the first CAs

watchCertAuthorities now compares all CAs it receives from the
watcher with the previous CA of the same type and only sends to
the remote site if they are not identical. This is to reduce
unnecessary network traffic which can be problematic for a
root cluster with a larger number of leafs. 

The CertAuthorityWatcher is refactored to leverage a fanout
to emit events to any number of watchers, each subscription
can be for a subset of the configured CA types. The proxy
now has only one CertAuthorityWatcher that is passed around
similarly to the LockWatcher. This reduces the memory usage
for proxies, which prior to this has one local CAWatcher per
remote site.

updateCertAuthorities no longer waits on the utils.Retry it
is provided with before starting to watch CAs. By doing this
the proxy no longer has to wait ~8 minutes before it even 
starts to watch CAs.

### CODE REVIEW

File: lib/services/watcher.go [hunk 1]
```diff
@@ -1009,28 +1065,23 @@ func (c *caCollector) resourceKind() string {
...
-	c.collectedCAs = updatedCerts
-	c.lock.Unlock()
+	defer c.lock.Unlock()
 
-	select {
-	case <-ctx.Done():
-		return trace.Wrap(ctx.Err())
-	case c.CertAuthorityC <- updatedCerts.ToSlice():
+	for _, ca := range cas {
+		c.cas[ca.GetType()][ca.GetName()] = ca
```
Comment (espadolini) [inline lib/services/watcher.go:1083]: Do you know if `GetCertAuthorities` is guaranteed to doublecheck the type of the CAs? If not we should probably doublecheck it here.

**Review by espadolini — APPROVED** [review]

File: lib/services/watcher.go [hunk 2]
```diff
@@ -961,13 +956,20 @@ func NewCertAuthorityWatcher(ctx context.Context, cfg CertAuthorityWatcherConfig
 
 	collector := &caCollector{
 		CertAuthorityWatcherConfig: cfg,
+		fanout:                     NewFanout(),
+		cas:                        make(map[types.CertAuthType]map[string]types.CertAuthority, len(cfg.Types)),
```
Comment (jakule) [inline lib/services/watcher.go:960]: nit:
```suggestion
		cas:                        make(types.CertAuthorityTypeMap, len(cfg.Types)),
```
I also don't mind renaming the `CertAuthorityTypeMap` if you have any better idea.

File: lib/services/watcher.go [hunk 3]
```diff
@@ -961,13 +956,20 @@ func NewCertAuthorityWatcher(ctx context.Context, cfg CertAuthorityWatcherConfig
...
+	}
+
+	for _, t := range cfg.Types {
+		collector.cas[t] = make(map[string]types.CertAuthority)
```
Comment (jakule) [inline lib/services/watcher.go:964]: nit:
```suggestion
		collector.cas[t] = make(CertAuthorityMap)
```

File: lib/services/watcher.go [hunk 4]
```diff
@@ -980,9 +982,63 @@ type CertAuthorityWatcher struct {
 // caCollector accompanies resourceWatcher when monitoring cert authority resources.
 type caCollector struct {
 	CertAuthorityWatcherConfig
+	lock   sync.RWMutex
+	cas    map[types.CertAuthType]map[string]types.CertAuthority
```
Comment (jakule) [inline lib/services/watcher.go:986]: nit:
```suggestion
	cas    CertAuthorityTypeMap
```

File: lib/services/watcher.go [hunk 5]
```diff
@@ -980,9 +982,63 @@ type CertAuthorityWatcher struct {
 // caCollector accompanies resourceWatcher when monitoring cert authority resources.
 type caCollector struct {
 	CertAuthorityWatcherConfig
+	lock   sync.RWMutex
```
Comment (jakule) [inline lib/services/watcher.go:985]: Can you add a comment or name the `lock` to indicate what are we locking with it? Whole struct or just one/some members.

File: lib/services/watcher.go [hunk 6]
```diff
@@ -1046,46 +1101,44 @@ func (c *caCollector) processEventAndUpdateCurrent(ctx context.Context, event ty
...
-			delete(c.collectedCAs[caType], event.Resource.GetName())
+		if !c.watchingType(caType) {
+			return
 		}
 
-		select {
-		case <-ctx.Done():
-		case c.CertAuthorityC <- c.collectedCAs.ToSlice():
-		}
+		delete(c.cas[caType], event.Resource.GetName())
```
Comment (jakule) [inline lib/services/watcher.go:1108]: As we modify `c.cas` shouldn't we use mutex here?

File: lib/services/watcher.go [hunk 7]
```diff
@@ -1046,46 +1101,44 @@ func (c *caCollector) processEventAndUpdateCurrent(ctx context.Context, event ty
...
-		select {
-		case <-ctx.Done():
-		case c.CertAuthorityC <- c.collectedCAs.ToSlice():
+		authority, ok := c.cas[ca.GetType()][ca.GetName()]
+		if ok && CertAuthoritiesEquivalent(authority, ca) {
+			return
 		}
+
+		c.cas[ca.GetType()][ca.GetName()] = ca
+		c.fanout.Emit(event)
```
Comment (jakule) [inline lib/services/watcher.go:1127]: As we read/modify `c.cas` shouldn't we use mutex here?

**Comment by jakule** [comment]
Unit test failure looks like something related:
```
    sshserver_test.go:1888: 
        	Error Trace:	sshserver_test.go:1888
        	Error:      	Received unexpected error:
        	            	missing parameter CertAuthorityWatcher
        	Test:       	TestIgnorePuTTYSimpleChannel
```

File: lib/services/watcher.go (hunk 2 above)
Comment (jakule) [reply lib/services/watcher.go:960]: I don't mind removing them, just `map[types.CertAuthType]map[string]types.CertAuthority` looks a bit "scary" in my opinion.

**Review by jakule — APPROVED** [review]

**Pull request merged**